### PR TITLE
Dx api

### PR DESCRIPTION
**Pull request by boa85**

### CODE REVIEW

File: src/xbridge/rpcxbridge.cpp [hunk 1]
```diff
@@ -315,12 +480,213 @@ Value dxCancelTransaction(const Array & params, bool fHelp)
...
+json_spirit::Value dxGetOrderBook(const json_spirit::Array& params, bool fHelp)
+{
+    if (fHelp || (params.size() != 3 && params.size() != 4))
+    {
+        throw runtime_error("dxGetOrderBook "
+                            "(the level of detail) (from currency) (to currency) (max orders - optional, default = 50) ");
+    }
+
+    Array arr;
+    boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
```
Comment (corvaxx) [inline src/xbridge/rpcxbridge.cpp]: don't hold the mutex for long read/sort/etc operations, pls

File: src/xbridge/util/xbridgeerror.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#include "xbridgeerror.h"
+namespace xbridge {
+    const std::string xbridgeErrorText(const Error error, const std::string &argumrnt)
+    {
+        switch (error)
+        {
+        case Error::INVALID_CURRENCY:
+            return "invalid currency " + argumrnt;
```
Comment (corvaxx) [inline src/xbridge/util/xbridgeerror.cpp]: and italian, spain, hindy and russian? :)
see _()

Comment (corvaxx) [reply src/xbridge/util/xbridgeerror.cpp]: or need to move this to gui if posible

File: src/xbridge/xbridge.cpp [hunk 3]
```diff
@@ -148,6 +148,7 @@ void XBridge::onTimer()
 {
     // DEBUG_TRACE();
 
+    LOG() << "xbridge::onTimer ";
```
Comment (corvaxx) [inline src/xbridge/xbridge.cpp]: debug logs? log files not an a garbage

Comment (corvaxx) [reply src/xbridge/xbridge.cpp]: and see DEBUG_TRACE

File: src/xbridge/xbridgeapp.cpp [hunk 4]
```diff
@@ -57,8 +60,20 @@ void badaboom()
...
+  , m_timer(m_timerIo, boost::posix_time::seconds(3))
 {
+    for (int i = 0; i < 2; ++i)
+    {
+        IoServicePtr ios(new boost::asio::io_service);
+
+        m_services.push_back(ios);
+        m_works.push_back(WorkPtr(new boost::asio::io_service::work(*ios)));
+        m_threads.create_thread(boost::bind(&boost::asio::io_service::run, ios));
+    }
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.cpp]: mister, we have more threads!
what does this?
and timer...

File: src/xbridge/xbridgeapp.cpp [hunk 5]
```diff
@@ -144,13 +160,28 @@ bool XBridgeApp::init(int argc, char *argv[])
     XBridgeExchange & e = XBridgeExchange::instance();
     e.init();
 
+    m_historicTransactionsStates = {XBridgeTransactionDescr::trExpired,
+                                    XBridgeTransactionDescr::trOffline,
+                                    XBridgeTransactionDescr::trFinished,
+                                    XBridgeTransactionDescr::trDropped,
+                                    XBridgeTransactionDescr::trCancelled,
+                                    XBridgeTransactionDescr::trInvalid};
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.cpp]: may be simple compare? not an a map? 
rNew <= state <= trFinished ?

File: src/xbridge/xbridgeapp.cpp [hunk 6]
```diff
@@ -315,6 +346,14 @@ void XBridgeApp::sleep(const unsigned int umilliseconds)
     boost::this_thread::sleep_for(boost::chrono::milliseconds(umilliseconds));
 }
 
+
+bool XBridgeApp::isHistoricState(const XBridgeTransactionDescr::State state)
+{
+    return std::find(m_historicTransactionsStates.begin(), m_historicTransactionsStates.end(), state) != m_historicTransactionsStates.end();
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.cpp]: it's a map

File: src/xbridge/xbridgeapp.h [hunk 7]
```diff
@@ -145,6 +149,50 @@ class XBridgeApp
...
+     */
+    boost::asio::deadline_timer m_timer;
+
+public:
+    /**
+     * @brief isHistoricState - checks the state of the transaction
+     * @param state - current state of transaction
+     * @return true, if the transaction is historical
+     */
+    bool isHistoricState(const XBridgeTransactionDescr::State state);
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.h]: (minor)

class name
public methods
public data
private methods
private data

File: src/xbridge/xbridgesession.cpp [hunk 8]
```diff
@@ -593,6 +593,7 @@ bool XBridgeSession::processTransactionAccepting(XBridgePacketPtr packet)
 //******************************************************************************
 bool XBridgeSession::processTransactionHold(XBridgePacketPtr packet)
 {
+    LOG() << "XBridgeSession::processTransactionHold " << __FUNCTION__;
```
Comment (corvaxx) [inline src/xbridge/xbridgesession.cpp]: debug logs?

File: src/xbridge/xbridgesession.cpp [hunk 9]
```diff
@@ -2247,6 +2248,7 @@ bool XBridgeSession::processTransactionConfirmedB(XBridgePacketPtr packet)
 //*****************************************************************************
 bool XBridgeSession::processTransactionCancel(XBridgePacketPtr packet)
 {
+    LOG() << "XBridgeSession::processTransactionCancel 2" << __FUNCTION__;
```
Comment (corvaxx) [inline src/xbridge/xbridgesession.cpp]: and this and more

File: src/rpcserver.cpp [hunk 10]
```diff
@@ -363,13 +363,15 @@ static const CRPCCommand vRPCCommands[] =
...
-        {"xbridge", "dxCancelTransaction",            &dxCancelTransaction,           true, true, true},
+        {"xbridge", "dxGetTransactions",                 &dxGetTransactions,                  true, true, true},
+        {"xbridge", "dxGetTransactionsHistory",         &dxGetTransactionsHistory,          true, true, true},
+        {"xbridge", "dxGetTransactionInfo",                 &dxGetTransactionInfo,                  true, true, true},
+        {"xbridge", "dxGetCurrencies",                    &dxGetCurrencies,                     true, true, true},
+        {"xbridge", "dxCreateTransaction",                  &dxCreateTransaction,                   true, true, true},
+        {"xbridge", "dxAcceptTransaction",                  &dxAcceptTransaction,                   true, true, true},
+        {"xbridge", "dxCancelTransaction",                  &dxCancelTransaction,                   true, true, true},
+        {"xbridge", "dxGetTradeHistory",                    &dxGetTradeHistory,    true, true, true},
+        {"xbridge", "dxGetOrderBook",  &dxGetOrderBook,    true, true, true},
```
[the remark this answers is not in the document]
Comment (kitovyj) [reply src/rpcserver.cpp]: Are you about the names or extra spaces?

[the remark this answers is not in the document]
Comment (corvaxx) [reply src/rpcserver.cpp]: spaces

File: src/xbridge/rpcxbridge.cpp [hunk 11]
```diff
@@ -143,33 +293,30 @@ Value dxGetTransactionInfo(const Array & params, bool fHelp)
...
     // pending tx
     {
-        std::map<uint256, XBridgeTransactionDescrPtr> trlist = XBridgeApp::m_pendingTransactions;
+        TransactionMap trlist = XBridgeApp::m_pendingTransactions;
         for (const auto & trEntry : trlist)
         {
             const auto tr = trEntry.second;
-
             if(id != tr->id.GetHex())
+            {
```
[the remark this answers is not in the document]
Comment (kitovyj) [reply src/xbridge/rpcxbridge.cpp]: Ops, epic fail) That is legacy code, though.

File: src/qt/xbridgeui/xbridgetransactionsmodel.cpp [hunk 12]
```diff
@@ -306,6 +310,43 @@ void XBridgeTransactionsModel::onTimer()
...
+                    LOG() << "can't remove from pending transaction, transaction " << id.GetHex() << " not found " << __FUNCTION__;
+                }
+            }
+        }
+        if(XBridgeApp::instance().isHistoricState(m_transactions[i].state))
+        {//add transaction into history
+            //@TODO replace model to smartpointers
+            XBridgeTransactionDescrPtr tmp = XBridgeTransactionDescrPtr(new XBridgeTransactionDescr(m_transactions[i]));
+            boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
+            XBridgeApp::m_historicTransactions[id] = tmp;
```
[the remark this answers is not in the document]
Comment (boa85) [reply src/qt/xbridgeui/xbridgetransactionsmodel.cpp]: it is necessary to bring the core and UI for a single model

File: src/xbridge/xbridgeapp.cpp (hunk 6 above)
Comment (boa85) [reply src/xbridge/xbridgeapp.cpp]: it's list
std::list<XBridgeTransactionDescr::State>       m_historicTransactionsStates;

Comment (corvaxx) [reply src/xbridge/xbridgeapp.cpp]: list...why not set or map? or simple compare?

File: src/xbridge/xbridge.cpp (hunk 3 above)
Comment (boa85) [reply src/xbridge/xbridge.cpp]: fix

File: src/qt/xbridgeui/xbridgetransactionsmodel.cpp [hunk 13]
```diff
@@ -306,6 +310,43 @@ void XBridgeTransactionsModel::onTimer()
             m_transactions.erase(m_transactions.begin()+i);
             emit endRemoveRows();
             --i;
+            {
+                boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
+                if(XBridgeApp::m_historicTransactions.find(id) != XBridgeApp::m_historicTransactions.end())
+                {
+                    XBridgeApp::m_historicTransactions.erase(id);
```
[the remark this answers is not in the document]
Comment (boa85) [reply src/qt/xbridgeui/xbridgetransactionsmodel.cpp]: fixed

File: src/qt/xbridgeui/xbridgetransactionsmodel.cpp [hunk 14]
```diff
@@ -306,6 +310,43 @@ void XBridgeTransactionsModel::onTimer()
...
+            LOG() << "insert into history transactions map " <<  m_transactions[i].strState() << "\t" << __FUNCTION__;
+            {
+                boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
+                XBridgeApp::m_historicTransactions[id] = tmp;
+            }
+            {
+                boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
+                if(XBridgeApp::m_pendingTransactions.find(id) != XBridgeApp::m_pendingTransactions.end())
+                {
+                    XBridgeApp::m_pendingTransactions.erase(id);
```
[the remark this answers is not in the document]
Comment (boa85) [reply src/qt/xbridgeui/xbridgetransactionsmodel.cpp]: fixed

File: src/rpcserver.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (boa85) [reply src/rpcserver.cpp]: fixed

File: src/xbridge/rpcxbridge.cpp [hunk 15]
```diff
@@ -42,86 +68,210 @@ Value dxGetTransactionList(const Array & params, bool fHelp)
...
-        throw runtime_error("dxGetTransactionsHistoryList\nHistoric list transactions.");
+        throw runtime_error("dxGetTransactionsHistory "
+                            "(ALL - optional parameter, if specified then all transactions are shown, "
+                            "not only successfully completed ");
     }
-
+    bool isShowAll = params.size() == 1 && params[0].get_str() == "ALL";
     Array arr;
-
     boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
```
[the remark this answers is not in the document]
Comment (boa85) [reply src/xbridge/rpcxbridge.cpp]: fixed

File: src/xbridge/rpcxbridge.cpp [hunk 16]
```diff
@@ -42,86 +68,210 @@ Value dxGetTransactionList(const Array & params, bool fHelp)
...
+{
+
+    if (fHelp || (params.size() != 4 && params.size() != 5))
+    {
+        throw runtime_error("dxGetTradeHistory "
+                            "(from currency) (to currency) (start time) (end time) (txids - optional) ");
+    }
 
+    Array arr;
+    boost::mutex::scoped_lock l(XBridgeApp::m_txLocker);
```
[the remark this answers is not in the document]
Comment (boa85) [reply src/xbridge/rpcxbridge.cpp]: fixed

File: src/xbridge/xbridgeapp.h (hunk 7 above)
Comment (boa85) [reply src/xbridge/xbridgeapp.h]: corrected

**Pull request merged**